README.md
# q2-karenina
[![Build Status](https://travis-ci.org/zaneveld/q2-karenina.svg?branch=master)](https://travis-ci.org/zaneveld/q2-karenina)

### Approach

q2-karenina in a QIIME2 plug-in for the karenina python package, which provides simulation and model-fitting tools for studying stochastic variability in microbial communities. The package is named for Anna Karenina Principle effects, in which certain diseases and environmental stressors can cause increased stochastic variability in animal microbiomes. A recent review of this phenomenon can be found in [Zaneveld et al.,Nature Microbiology, 2017](https://www.ncbi.nlm.nih.gov/pubmed/28836573). This pattern is in contrast to stressors that shift the mean *position* of microbial communities in PCoA space.

The approach of the q2-karenina package to modeling microbiome dynamics is top down: microbial communities in PCoA space are modeled as if they were physical particle in Euclidean space. For example, a microbiome that shifts in composition by a small amount at each time-step can be modeled like the random walk of a particle undergoing Brownian Motion. Most healthy animal microbiomes do not appear to wander randomly like this. Instead, they tend to revert over time to a composition that is average for a particular individual. q2-karenina models this phenomenon using Ornstein-Uhlenbeck models, which include both Brownian motion and reversion to a particular home position. 

### Model

This model is fairly simple, having only 3 tunable parameters per PC axes. (A separate model is fit to the  motion of a microbial community on each PC axis separately, as many papers observe increased variance along some axes but not others).

Mathematically, the model is as follows:

dx/dt = dW * sigma + (x - theta)* lambda

**Where**:

dx/dt = change in the position of a microbial community on a particular axis of the PCoA plot 

dW = the Weiner process which generates Brownian motion (effectively drawing the extent of displacement from a normal distribution)

sigma = a constant that scales the Brownian motion component

theta = the 'home position' or attractor for the community. Over time (in the absence of disturbance or e.g. seasonal effects that alter the mean composition) it is expected that this position will be the mean composition of microbiome samples from an individual. It is also expected that in a stable community this will be the centroid of samples taken from an individual longitudinally.

### Release

The initial q2-karenina release allows for fitting of Ornstein-Uhlenbeck models to QIIME2 ordination results using the fit_spatial_ornstein_uhlenbeck visualizer, and returns the parameters for the model fit. Future development will wrap other functionality in karenina such as simulation of QIIME2-compatible OrdinationResult and Metadata files based on Ornstein-Uhlenbeck model parameters, benchmarks of the model fitting procedure, and visualization.

**Tutorial**
* [Tutorial](https://github.com/zaneveld/q2-karenina/blob/master/doc/q2-karenina_tutorial.md)

**Other links**
* [karenina](https://github.com/zaneveld/karenina)
* [Documentation](https://zaneveld.github.io/karenina/html/index.html)
* [QIIME 2](https://qiime2.org)


Models can either be fit to timeseries from each subject separately (individual fit), or can be fit to all individuals within a treatment collectively (cohort fit) using the spatial_ornstein_uhlenbeck.py script




CHANGELOG.md
# q2-karenina changelog

9/24/2018

-- set dependency to www.github.com/zaneveld/karenina
-- added output of index files to qzv (these are required to display properly)
-- updated setup.py to make these accessible from pkg_resources
-- removed duplicative code in _fit_timeseries (rely on the canonical version in karenina)


q2_karenina/_visualization.py
#/usr/bin/env python
from __future__ import division

__author__ = "Jesse Zaneveld"
__copyright__ = "Copyright 2016, The Karenina Project"
__credits__ = ["Jesse Zaneveld"]
__license__ = "GPL"
__version__ = "0.0.1-dev"
__maintainer__ = "Jesse Zaneveld"
__email__ = "zaneveld@gmail.com"
__status__ = "Development"

from karenina.individual import Individual
import karenina.visualization as k_visualization
import pkg_resources
import qiime2
import q2templates
from q2_types.ordination import PCoAResults
import pandas as pd

def visualization(output_dir: str, pcoa : str, metadata : str,
				 individual_col: str, timepoint_col: str, treatment_col: str):
    """
	Connect to karenina.visualization and save output
	
	:param output_dir: output directory
	:param pcoa: pcoa qza file
	:param metadata: metadata file location
	:param individual_col: individual column identifiers
	:param timepoint_col: timepoint column identifier
	:param treatment_col: treatment column identifier
	"""
	
	# Parse in pcoa and metadata as dataframes and inject to k_visualization
    ind = []
    site = _parse_pcoa(pcoa)
    df = _parse_metadata(metadata, individual_col, timepoint_col, treatment_col, site)
    i = 0
    colors = ['fuchsia', 'cyan', 'darkorange', 'blue', 'yellow']
    tx = treatment_col
    treatments = df[tx].unique()
    while len(colors) < len(treatments):
	    colors.append('lightgray')

    for row in df.iterrows():
        curr_subject_id = "%s_%i" % (df[individual_col], i)
        j = 0
        while row[1][3] != treatments[j]:
            j += 1
        color = colors[j]
        params = {'lambda': 0.2, 'delta': 0.25, 'interindividual_variation': 0.01}
        params['color'] = color
        curr_subject = Individual(subject_id = curr_subject_id,
                                  params = params, \
                                  metadata = {treatment_col: df[treatment_col]}, \
                                  interindividual_variation = .01)
        ind.append(curr_subject)
        i += 1

    k_visualization.save_simulation_figure(individuals = ind, 
	                output_folder = output_dir, n_timepoints = 50, 
	                perturbation_timepoint = 25, n_individuals = 50)
    k_visualization.save_simulation_movie(individuals = ind,
	                output_folder = output_dir, n_timepoints = 50, n_individuals = 50)
	
def _parse_pcoa(pcoa):
    # Parse PCoA Contents
    pcoa = open(pcoa,"rb")
    lines = pcoa.readlines()
    site = []
    i = 0
    for line in lines:
        line = line.decode("utf-8")
        if line.startswith("Eigvals"):
            eigs = lines[i+1].decode("utf-8").strip('\n').strip('\r').split("\t")
        elif line.startswith("Proportion explained"):
            propEx = lines[i+1].decode("utf-8").strip('\n').strip('\r').split("\t")
        elif line.startswith("Species"):
            species = lines[i + 1].decode("utf-8").strip('\n').strip('\r').split("\t")

        elif line.startswith("Site"):
            # We don't want Site constraints.
            if ("constraints") in line:
                break
            max = int(line.split("\t")[1])+i
            j = i + 1
            while j <= max:
                t_line = lines[j].decode('utf-8')
                site.append(t_line.strip('\n').strip('\r').split("\t"))
                j += 1
        i += 1
    t_site = []
    for item in site:
        t_site.append([item[0],item[1],item[2],item[3]])
    site = t_site

    #We now have variables:
            #First three are stored for now, will be utilized later when standardization of axes can be supported.
        # eigs: every eigenvalue
        # propEx: every proportion explained
        # species: every species

        # site: Every site, with the highest three proportion-explained PCs, in the form:
            # subjectID, x1, x2, x3
    return site

def _parse_metadata(metadata, individual_col, timepoint_col, treatment_col, site):
    df = pd.read_csv(metadata,sep="\t")
    # Drop any rows that are informational
    while df.iloc[0][0].startswith("#"):
	    df.drop(df.index[:1], inplace=True)
	
    # Combine Individual columns if multiple subject identifiers defined (Such as individual and site)
    if "," in individual_col:
        individual_temp = individual_col.split(",")
        df_ind = df[individual_temp[0]]

        # Iterate over the columns and combine on each pass
        for i in range(len(individual_temp)):
            if i == 0:
                pass
            else:
                df_ind = pd.concat([df_ind,df[individual_temp[i]]],axis=1)

        # Hyphenate the white space between identifiers
        for i in range(len(individual_temp)):
            if i == 0:
                pass
            else:
                df_ind = df_ind[individual_temp[0]].astype(str)+"_"+df_ind[individual_temp[1]]

        # Remove any user-generated white space
        inds = []
        for row in df_ind.iteritems():
            inds.append(row[1].replace(" ","-"))
        df_ind = pd.DataFrame({individual_col:inds})
        # Reindex to match dataframes to be merged
        df_ind.index += 1
    else:
        df_ind = df[individual_col]

    df_tp = df[timepoint_col]
    
    #Copy individual column into treatment if none assigned
    if treatment_col is not None:
        df_tx = df[treatment_col]
    else:
        df_tx = df_ind.copy()
        df_tx.columns = [str(df_tx.columns.values[0])+"_tx"]

    # Force timepoint to numerics
    df_tp = df_tp.replace('[^0-9]','', regex=True)


    # Build the return dataframe
    df_ret = pd.concat([df[df.columns[0]], df_ind, df_tp, df_tx], axis=1)

    df_site = pd.DataFrame.from_records(site, columns=["#SampleID","pc1","pc2","pc3"])

    df_ret = pd.merge(df_ret, df_site, on="#SampleID", how='inner')

    if "," in individual_col:
        df_ret.rename(columns = {0:str(individual)}, inplace=True)

    #Now have a full dataframe with sampleIDs, Subject Identifiers, timepoints, treatment, and initial x coordinates.
        #Preserves only values which have initial positions, drops null coordinates.
        #In example files, L3S242 was in metadata File, but not in the PCOA file, so there was one less in the df_ret.
    return df_ret

q2_karenina/_fit_timeseries.py
#/usr/bin/env python
from __future__ import division

__author__ = "Jesse Zaneveld"
__copyright__ = "Copyright 2016, The Karenina Project"
__credits__ = ["Jesse Zaneveld","Samuel L. Peoples"]
__license__ = "GPL"
__version__ = "0.1-dev"
__maintainer__ = "Jesse Zaneveld"
__email__ = "zaneveld@gmail.com"
__status__ = "Development"

from q2templates import render
from os.path import join
from pkg_resources import resource_filename
from karenina.fit_timeseries import parse_pcoa,parse_metadata,fit_input

def fit_timeseries(output_dir: str, pcoa : str, metadata:str, method : str, 
                individual_col: str, timepoint_col: str, treatment_col: str) -> None:
    
    #Handle missing parameters from user interface
    if 'None' in metadata:
	    metadata = None
    if 'None' in treatment_col:
	    treatment_col = None

    #read in and format input data     
    site, metadata = parse_pcoa(pcoa, individual_col,\
      timepoint_col, treatment_col, metadata=None)
    model_input = parse_metadata(metadata, individual_col,\
      timepoint_col, treatment_col, site)
   
     
    if treatment_col is not None:
        #If there is a non-empty treatment column, then fit a cohort model,
        # in which all individuals with the same treatment
        #have the same model parameters.
        
        output, cohort_output = fit_input(model_input,\
          individual_col, timepoint_col, treatment_col, method)
        
        result = cohort_output
        plot_name = 'Ornstein-Uhlenbeck cohort model fit results (treatment = {},method={})'.format(treatment_col,method)
    else:
        #IF no treatment is provided, fit a model to each individual
 
       result = fit_input(model_input, individual_col,\
          timepoint_col, treatment_col, method)
 
       plot_name = 'Ornstein-Uhlenbeck individual model fit results (method={})'.format(method)
   
    #Output the results (individual or cohort based model fit) to a CSV filte 
    result.to_csv(join(output_dir,"fit_timeseries_results.csv"), index=False)

    #Generate the index.html file required by QIIME2 by filling in the template
    render_index_html(output_dir,plot_name)


def render_index_html(output_dir,plot_name):

    #Find the filepath for the q2_emperor folder 'assets'.
    #NOTE: to understand what's happening here see e.g. stackoverflow example here:
    #https://stackoverflow.com/questions/39104/finding-a-file-in-a-python-module-distribution

    template_dir = resource_filename('q2_karenina', 'assets')

    #get the path to our basic, unfilled index.html file (in the assets folder of q2_emperor)
    index = join(template_dir, 'index.html')

    #Use q2_templates.render to fill in data specific to our output in this visualization.
    #Documentation for q2_templates.render is available here:
    # https://github.com/qiime2/q2templates/blob/master/q2templates/_templates.py

    render(index, output_dir, context={'plot_name': plot_name})





q2_karenina/_spatial_ornstein_uhlenbeck.py
#/usr/bin/env python
from __future__ import division

__author__ = "Jesse Zaneveld"
__copyright__ = "Copyright 2016, The Karenina Project"
__credits__ = ["Jesse Zaneveld"]
__license__ = "GPL"
__version__ = "0.0.1-dev"
__maintainer__ = "Jesse Zaneveld"
__email__ = "zaneveld@gmail.com"
__status__ = "Development"

import karenina.spatial_ornstein_uhlenbeck as k_OU
from karenina.experiment import Experiment
import pkg_resources
import qiime2
import q2templates
from scipy.spatial import distance
from skbio import OrdinationResults
from q2_types.ordination import PCoAResults
from q2_types.ordination import PCoAResults
from q2_types.distance_matrix import DistanceMatrix
import pandas as pd
import tempfile
import os
        
def spatial_ornstein_uhlenbeck(perturbation_fp:str, treatment_names:str, n_individuals:str, 
                            n_timepoints:int, perturbation_timepoint:int,
                            perturbation_duration:int,interindividual_variation:float,
                            delta:float,lam:float,fixed_start_pos:str) -> (OrdinationResults, DistanceMatrix):
    #pass items to k_OU, 
    #ensure that metadata is saved with both PcOAResults, and DistanceMatrix!
    k_OU.check_perturbation_timepoint(perturbation_timepoint,n_timepoints)
    individual_base_params = {"lambda":lam,"delta":delta,
        "interindividual_variation":interindividual_variation}
    if "None" in fixed_start_pos:
        fixed_start_pos = None
    if fixed_start_pos:
        try:
            x,y,z = map(float,fixed_start_pos.split(","))
            individual_base_params['x']=x
            individual_base_params['y']=y
            individual_base_params['z']=z

        except:
            print ("Supplied value for fixed start position after parsing:",fixed_start_pos)
            raise ValueError('Problem with --fixed_start_pos. Got %s Please supply x,y,z values in the range (-1,1) separated by commas and '+
                'enclosed in quotes. Example: "0.1,-0.2,0.3"'% fixed_start_pos)
    perturbations = k_OU.parse_perturbation_file(perturbation_fp, perturbation_timepoint, perturbation_duration)
    treatments = [[], perturbations]
    treatment_names = treatment_names.split(",")
    n_individuals = list(map(int,n_individuals.split(",")))
    experiment = Experiment(treatment_names,n_individuals,n_timepoints,
        individual_base_params,treatments,interindividual_variation, verbose=False)
    experiment.simulate_timesteps(0,n_timepoints, verbose=False)
    data, ids = experiment.q2_data()
    return _simulation_data(data, ids)
    
def _simulation_data(data, ids):
    with open("ordination.txt","w", encoding='utf8') as ordination:
        ordination.write('Eigvals\t0'+'\n\n')
        ordination.write('Proportion explained\t0'+'\n\n')
        ordination.write('Species\t0\t0\n\n')
        ordination.write('Site\t'+str(len(data)*len(data[0][0]))+'\t3\n')
        dm = {}
        j=0
        for row in data:
            identifier = ids[j]
            for i in range(len(row[0])):
                ordination.write(str(identifier)+"_t"+str(i)+"\t"+str(row[0][i])+"\t"+str(row[1][i])+"\t"+str(row[2][i])+"\n")
                dm.update({str(identifier)+"."+str(i):[row[0][i],row[1][i],row[2][i]]})
            j+=1
        ordination.write("\n")
        ordination.write("Biplot\t0\t0\n\n")
        ordination.write("Site constraints\t0\t0\n")
        ordination_results = OrdinationResults.read("ordination.txt")
    ordination.close
    os.remove("ordination.txt")
    
    # Distance matrix (euclidean)
    dm_0 = []
    dm_0.append("")
    distance_matrix = []
    for key in dm.keys():
        dm_0.append(key)
    distance_matrix.append(dm_0)
    for key in dm.keys():
        dm_1 = []
        dm_1.append(key)
        for key1 in dm.keys():
            dm_1.append(str(distance.euclidean(dm[key],dm[key1])))
        distance_matrix.append(dm_1)

    #Mapping file
    md_0 = ["#SampleID","Subject","Treatment","Timepoint"]
    md_1 = ["#q2:types","categorical","categorical","numeric"]
    md = []
    for id in ids:
        for i in range(len(data[0][0])):
            md.append([id+"_t"+str(i),id,''.join([k for k in id if not k.isdigit()])[:-1],i])
    metadata = [md_0,md_1]
    for row in md:
        metadata.append(row)
    #ADD FUNCTIONALITY TO RETURN MAPPING FILE
    return ordination_results, distance_matrix

q2_karenina/plugin_setup.py
import qiime2.plugin

import q2_karenina
from q2_karenina._spatial_ornstein_uhlenbeck import spatial_ornstein_uhlenbeck
from q2_karenina._fit_timeseries import fit_timeseries
from q2_karenina._visualization import visualization
from q2_types.ordination import PCoAResults
from q2_types.distance_matrix import DistanceMatrix
from qiime2.plugin import Metadata, Str, Choices, Int, Float


plugin = qiime2.plugin.Plugin(
    name='karenina',
    version=q2_karenina.__version__,
    website='https://github.com/zaneveld/karenina',
    package='q2_karenina',
    user_support_text=None,
    description="This script simulates microbiome " +
    "change over time using Ornstein-Uhlenbeck (OU) models.  These are " +
    "similar to Brownian motion models, with the exception that they " +
    "include reversion to a mean. Output is a tab-delimited data table " +
    "and figures.",
    citation_text=None
)
""""""
# Method omitted temporarily until:
    # Output can be in qiime2 Metadata, PcOAResults, and DistanceMatrix format

plugin.methods.register_function(
    function=spatial_ornstein_uhlenbeck,
    inputs = {},
    parameters = {
        'perturbation_fp': Str,
        'treatment_names': Str,
        'n_individuals': Str,
        'n_timepoints': Int,
        'perturbation_timepoint': Int,
        'perturbation_duration': Int,
        'interindividual_variation': Float,
        'delta': Float,
        'lam': Float,
        'fixed_start_pos': Str
    },
    outputs = [
        ('ordination', PCoAResults),
        ('distance_matrix', DistanceMatrix)
    ],
    input_descriptions = {},
    parameter_descriptions = {
        'perturbation_fp': 'filepath for perturbation parameters for simulation results',
        'treatment_names': '[\'control,destabilizing_treatment\'] Names for simulation treatments',
        'n_individuals': '[\'35,35\'] Number of individuals per treatment',
        'n_timepoints': '[\'10\'] Number of simulation timepoints',
        'perturbation_timepoint': '[\'5\'] Timepoint at which to apply treatment (<n_timepoints)',
        'perturbation_duration': '[\'100\'] Duration of perturbation.',
        'interindividual_variation': '[\'0.01\']Starting variability between individuals',
        'delta': '[\'0.25\'] Starting Delta parameter for Brownian Motion/ OU models. '
            +'Higher values indicate greater variability over time',
        'lam': '[\'0.20\'] Starting Lambda value for OU process. '
            +'Higher values indicate a greater tendancy to revert to the mean value.',
        'fixed_start_pos': 'Starting x,y,z position for each point. '
            +'If not defined, starting positions will be randomized based on '
            +'interindividual_variation; type: string, eg: [\'0.0,0.1,0.2\'].'
    },
    output_descriptions = {
        'ordination': 'Sample PCoA file containing simulation data',
        'distance_matrix': 'Sample Distance Matrix containing simulation data'
    },
    name = 'Spatial Ornstein Uhlenbeck microbial community simulation',
    description = ('This method simulates microbial behavior over time using'
                +'Ornstein Uhlenbeck models. This are similar to Brownian Motion'
                +'with the exception that they include reversion to a mean.')
)
""""""
    
# Modify to allow for PCoAResults as input    
plugin.visualizers.register_function(
    function = fit_timeseries,
    inputs = {
        #'pcoa' : PCoAResults
    },
    parameters = {
        'pcoa': Str,
        'method': Str % Choices({'basinhopping'}),
        'metadata': Str,
        'individual_col': Str,
        'timepoint_col': Str,
        'treatment_col': Str
    },
    parameter_descriptions = {
        'pcoa': 'filepath to PCoA results',
        'method': 'global optimization method',
        'metadata': 'filepath to Sample metadata',
        'individual_col': 'individual column identifier',
        'timepoint_col':'timepoint column identifier',
        'treatment_col': 'treatment column identifier'
    },
    name = 'Fit OU Models to PCoA Ordination output',
    description = 'This visualizer generates OU model parameters for PCoA output'
                'data, for each individual and each defined treatment cohort.'
)

plugin.visualizers.register_function(
    function = visualization,
    inputs = {
        #'pcoa' : PCoAResults
    },
    parameters = {
        'pcoa': Str,
        'metadata': Str,
        'individual_col': Str,
        'timepoint_col': Str,
        'treatment_col': Str
    },
    parameter_descriptions = {
        'pcoa': 'filepath to PCoA results',
        'metadata': 'filepath to Sample metadata',
        'individual_col': 'individual column identifier',
        'timepoint_col': 'timepoint column identifier',
        'treatment_col': 'treatment column identifier'
    },
    name = 'Generates 3D animations of PCoA Timeseries',
    description = 'This visualizer generates 3D animations of PCoA Timeseries.'
)
